indico/modules/rb_new/client/js/components/RoomSelector.jsx

```diff
@@ -0,0 +1,191 @@
+        initialValue: PropTypes.array
+    };
+
+    static defaultProps = {
+        initialValue: []
+    };
+
+    constructor(props) {
+        super(props);
+        const {initialValue, onChange} = this.props;
+
+        if (initialValue) {
+            onChange(initialValue);
+        }
+
+        this.state = {
+            isFetchingLocations: false,
+            locations: [],
+            selectedLocation: null,
+            selectedRoom: null,
+            selectedRooms: initialValue
+        };
+    }
+
+    removeItem = (room) => {
+        const {selectedRooms} = this.state;
+        const {onChange} = this.props;
+        const newRooms = selectedRooms.filter((selectedRoom) => selectedRoom.id !== room.id);
+
+        this.setState({selectedRooms: newRooms}, () => {
+            onChange(newRooms);
+        });
+    };
+
+    renderRoomItem = (room) => {
+        return (
+            <List.Item key={room.id}>
+                <Image src={room.large_photo_url} size="mini" />
+                <List.Content>
+                    <List.Header>{room.full_name}</List.Header>
+                </List.Content>
+                <List.Content styleName="remove-btn-content">
+                    <Icon name="remove"
+                          onClick={() => this.removeItem(room)} />
+                </List.Content>
+            </List.Item>
+        );
+    };
+
+    render() {
+        const {onChange} = this.props;
+        const {locations, selectedLocation, selectedRoom, selectedRooms: rooms, isFetchingLocations} = this.state;
+        const locationOptions = locations
+            .sort((a, b) => a.name.localeCompare(b.name))
+            .map((location) => {
+                return {text: location.name, value: location.id};
+            });
+        let roomOptions = [];
+        if (selectedLocation) {
+            roomOptions = selectedLocation.rooms
+                .filter((room) => rooms.findIndex((item) => item.id === room.id) === -1)
+                .sort((a, b) => a.full_name.localeCompare(b.full_name))
+                .map((room) => {
+                    return {text: room.full_name, value: room.id};
+                });
+        }
+
+        return (
+            <Grid styleName="room-selector">
+                <Grid.Row>
+                    <Grid.Column width={5}>
+                        <Dropdown placeholder={Translate.string('Location')}
+                                  options={locationOptions}
+                                  value={selectedLocation && selectedLocation.id}
+                                  loading={isFetchingLocations}
+                                  onChange={(event, data) => {
+                                      const selected = locations.find((loc) => loc.id === data.value);
+                                      this.setState({selectedLocation: selected});
+                                  }}
+                                  onOpen={() => {
```

Comment:
We're only loading the list of locations once the dropdown is opened. That results in a delay after clicking it. Since it seems quite cheap as a request, what about loading it in `componentDidMount(...)`?

Reply:
ahh, no.. I forgot I did that initially, however React would complain about setting the state in the `componentDidMount` method and that is why I moved it to the `onOpen` callback